### PR TITLE
Feature - `ArrayDataset`

### PR DESCRIPTION
**Pull request by Ce11an**
## What does this PR do?

Part of #839
- `pl_bolts.datamodules.sklearn_datamodule.TensorDataset`

## Summary
- Replaced `TensorDataset` with `ArrayDataset`. (breaking change ❗ )
- Moved `ArrayDataset` to `datasets` module.

The `ArrayDataset` allows you to input any number of `ARRAYS` along with callable transforms using the `DataModel`.  The `ArrayDataset` allows for the following types:
- Lists / Nested Lists
- Numpy Arrays
- Torch Tensors

This eliminates the need for the `TensorDataset`.

Transforms are called during each `__getitem__`. @otaj and I have discussed the potential inefficiencies of this. However, we have decided to stay consistent with the rest of `bolts` and `vision`. Happily open to having that discussion again, though!

Also, after discussing with @otaj, we thought it would be best to separate #846.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does **only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [x] Did you write any **new necessary tests**? \[not needed for typos/docs\]
- [x] Did you verify new and **existing tests** pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-bolts/blob/master/CHANGELOG.md)?

## PR review

- [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Of course 🥳

### CODE REVIEW

File: pl_bolts/datasets/base_dataset.py [hunk 1]
```diff
@@ -58,3 +60,31 @@ def _download_from_url(self, base_url: str, data_folder: str, file_name: str):
...
+
+    Ties together data and callable transforms.
+
+    Attributes:
+        data: Sequence of indexables.
+        transform: Callable to transform data.
+    """
+
+    data: ARRAYS
+    transform: Optional[Callable] = None
```
Comment (otaj) [inline pl_bolts/datasets/base_dataset.py:77]: The `transform` can be better typed:

```suggestion
    transform: Optional[Callable[[ARRAYS], ARRAYS]] = None
```

File: pl_bolts/utils/types.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from typing import List, Union
+
+import numpy as np
+import torch
+
+ARRAYS = Union[torch.Tensor, np.ndarray, List[Union[float, int]], List[List[Union[float, int]]]]
```
[the remark this answers is not in the document]
Comment (otaj) [reply pl_bolts/utils/types.py:6]: Btw, I'd vote for having the type with CamelCase and not with UPPERCASE

[the remark this answers is not in the document]
Comment (otaj) [reply pl_bolts/utils/types.py:6]: And maybe prepended with T to make it obvious that it's a type. I.e. resulting in something like `TArrays`

File: tests/datasets/test_array_dataset.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
+import numpy as np
+import pytest
+import torch
+from pytorch_lightning.utilities import exceptions
+
+from pl_bolts.datasets import ArrayDataset, DataModel
+from pl_bolts.datasets.utils import to_tensor
+
+
+def add_one(integers: np.ndarray) -> np.ndarray:
```
Comment (otaj) [inline tests/datasets/test_array_dataset.py:10]: Type it with `ARRAYS`, since it can accept anything.

Comment (otaj) [reply tests/datasets/test_array_dataset.py:10]: And it becomes a bit tricky, if it should accept all three types of `ARRAYS` (Tensors, NumPy Arrays and regular lists). You could maybe play with something like `functools.single_dispatch`

Comment (otaj) [reply tests/datasets/test_array_dataset.py:10]: Btw, these comments apply to all these auxiliary functions for testing

Comment (Ce11an) [reply tests/datasets/test_array_dataset.py:10]: I have decided that it is just easier to use the `to_tensor` function for testing 😅

File: pl_bolts/datasets/utils.py [hunk 4]
```diff
@@ -39,3 +42,16 @@ def prepare_sr_datasets(dataset: str, scale_factor: int, data_dir: str):
         dataset_test = dataset_cls(scale_factor, root=data_dir, split="test", download=True)
 
     return (dataset_train, dataset_val, dataset_test)
+
+
+def to_tensor(integers: List[int]) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (Ce11an) [reply pl_bolts/datasets/utils.py:47]: Using `TArrays`

File: pl_bolts/datasets/array_dataset.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    def __getitem__(self, idx: int) -> Tuple[ARRAYS, ...]:
+        return tuple(data_model.process(data_model.data[idx]) for data_model in self.data_models)
+
+    def _equal_size(self) -> bool:
+        """Checks the size of the data_models are equal in the first dimension.
+
+        Returns:
+            bool: True if size of data_models are equal in the first dimension. False, if not.
+        """
+        return all(len(data_model.data) == len(self.data_models[0].data) for data_model in self.data_models)
```
[the remark this answers is not in the document]
Comment (Ce11an) [reply pl_bolts/datasets/array_dataset.py:52]: Ooo very nice!

**Comment by Ce11an** [comment]
Hey 👋🏻

Sorry, I think someone needs to approve the workflows again. I have resolved @otaj's comments. Thanks 😃

**Comment by Ce11an** [comment]
Thanks, @otaj!

`mypy` does not support recursive types. Currently, there is an experimental feature: https://github.com/python/mypy/issues/731

What would you recommend?

**Comment by otaj** [comment]
> Thanks, @otaj!
> 
> `mypy` does not support recursive types. Currently, there is an experimental feature: [python/mypy#731](https://github.com/python/mypy/issues/731)
> 
> What would you recommend?

Yeah, I sadly know about it, `mypy` can sometimes be a bit of a pain. I don't know if `mypy` has problem only with definition line or with all the lines where the recursive type is being used. If it's the latter, then we probably need to abandon recursive type, since it doesn't make sense to ignore *everything*. But, if it's only the former case, then I'm perfectly fine with just `type: ignore` on that line

**Comment by Ce11an** [comment]
There were two issues. I have had to add `float` type and ignore `TArrays` in `mypy`.

Are you happy to use `float` to represent `int`? [Pep 484](https://peps.python.org/pep-0484/#the-numeric-tower:~:text=Rather%20than%20requiring,are%20exceedingly%20rare.)

**Review by otaj — APPROVED** [review]

**Pull request merged**